### PR TITLE
Tensor<rank,dim,Number> - major cleanup

### PR DESCRIPTION
**Pull request by tamiko**
In reference of #1433

This is a major cleanup of the Tensor<rank,dim,Number> implementation. All I
wanted was mixed scalar type operations between tensors and I stumbled upon a
lot of bogus code that piled up over the years. Instead of just slabbing
another round of features upon it, it radically rewrote the whole thing...

I'm halfway through. I will radically remove the numerous implementations of
```contract``` and ```operator*``` in a forthcoming pull-request (that will, as
a side-effect, also support mixed scalar type operations...).

Just for the record:
```
4 files changed, 710 insertions(+), 1611 deletions(-)
```
while adding functionality.

The biggest change is that I purged the specialization ```Tensor<1,dim,Number>```
entirely. It is sufficient to have _one_ specialization to ground the recursive
template pattern and ```Tensor<0,dim,Number>``` is the canonical candidate for
that. Furthermore, ```Tensor<1,dim,Number>``` and ```Tensor<rank,dim,Number>```
already shared 95% of the _same_ code.

Furthermore, I removed as many of the external function declarations as
possible from ```tensor_base.h``` and put them into ```tensor.h```.

An even more radical approach would be to decouple ```Point<dim,Number>``` from
```Tensor<dim,Number``` entirely - they do not share much code anyway...

I open this bug report while I will do a final full testsuite run over night
(and a build test with C++03) to make sure that I caught every corner case.
Reports tomorrow. Second pull-request with the other changes later this week.

Notable commits:

c709224 (Matthias Maier, 76 minutes ago)
   Minor Bugfixes on Point<dim>

   - use numbers::NumberTraits<Number>::abs_square to compute the squared
    euclidian norm

   - allow scaling with arbitrary, compatible number type

   - implement some functions directly to be independent of the Tensor<>
    implementation

eba5060 (Matthias Maier, 2 hours ago)
   Restructure declarations in tensor_base.h and tensor.h

   Move everything that isn't necessary from tensor_base.h to tensor.h

146fd25 (Matthias Maier, 20 hours ago)
   Eliminate specialization for Tensor<1,dim>

   This is more or less an exact code copy of Tensor<rank_,dim>.

   tensor_base.h now contains the full definition of Tensor<rank_,dim>. All
   external functions operating on tensors that are not strictly needed reside
   in tensor.h.

85eda3c (Matthias Maier, 25 hours ago)
   Tensor<k,dim>: Remove superfluous operator* variants

   It is completely sufficient to provide 1 ("one") templated variant.

   I guess, over time more and more variants were added. This increased the
   grade of ambiguity for the compiler to chose the one intended
   implementation.

8b4ed36 (Matthias Maier, 58 seconds ago)
   Intermediate: Clarify our intent

   This commit restructures the usage of the Tensor<rank,dim,Number> class a
   bit to make a reimplementation possible. It especially eliminates the
   following corner cases

   - use floating point number where floating point numbers are meant, e.g.,
    a * 4 versus a * 4.
    where a is a floating point object.

     This doesn't mean that we don't support implicit casts any more - it just
    makes porting easier...

   - Include <deal.II/base/tensor.h> where we intend to print a tensor. This
    allows to move just about every external operator declaration/definition
    from tensor_base.h to tensor.h

### CODE REVIEW

**Comment by tamiko** [comment]
Full testsuite run for C++11 [1]. I have to refactor some parts for C++98/03...

[1] http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=12004

**Comment by kronbichler** [comment]
I quickly looked into the changes and they look good. This is a very welcome patch (in particular with respect to functionality with VectorizedArray).

File: doc/news/changes.h [hunk 1]
```diff
@@ -114,6 +114,13 @@ inconvenience this causes.
 
 
 <ol>
+  <li> Cleanup: The interface of Tensor<rank,dim,Number> has been cleaned
+  up (A lot of unnecessary partial template specializations have been
```
Comment (bangerth) [inline doc/news/changes.h]: A -> a

File: doc/news/changes.h [hunk 2]
```diff
@@ -143,9 +150,10 @@ inconvenience this causes.
   - TensorFunction
 
   <br>
-  <em>Classes now supporting complex number types via header:</em>
+  <em>Classes now fully supporting complex number types via header:</em>
```
Comment (bangerth) [inline doc/news/changes.h]: can you elaborate what "via header" means? :-)

File: include/deal.II/base/point.h [hunk 3]
```diff
@@ -419,7 +443,7 @@ Point<dim,Number>::square () const
 {
   Number q = Number();
   for (unsigned int i=0; i<dim; ++i)
-    q += this->values[i] * this->values[i];
+    q += numbers::NumberTraits<Number>::abs_square(this->values[i]);
```
Comment (bangerth) [inline include/deal.II/base/point.h]: That's a change of behavior for complex data types. Can you add a note to the documentation of the function how it treats different `Number` types? In other words, how it interprets the term `square`.

File: include/deal.II/base/tensor.h [hunk 4]
```diff
@@ -19,741 +19,213 @@
...
+ * Output operator for tensors of rank 1. Print the elements consecutively,
+ * with a space in between.
+ *
+ * @relates Tensor<1,dim,Number>
+ */
+template <int dim, typename Number>
 inline
-typename Tensor<rank_,dim,Number>::real_type
-Tensor<rank_,dim,Number>::norm_square () const
+std::ostream &operator << (std::ostream &out, const Tensor<1,dim,Number> &p)
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: do you actually still need this? It looks like it'd be subsumed in the general case for arbitrary rank above.

File: include/deal.II/base/tensor.h [hunk 5]
```diff
@@ -19,741 +19,213 @@
...
+ * Output operator for tensors of rank 1 and dimension 1. This is implemented
+ * specialized from the general template in order to avoid a compiler warning
+ * that the loop is empty.
+ *
+ * @relates Tensor<1,dim,Number>
+ */
 inline
-void
-Tensor<rank_, dim, Number>::unroll (Vector<Number2> &result) const
+std::ostream &operator << (std::ostream &out, const Tensor<1,1,double> &p)
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: same here -- I think the general template above should do, no?

File: include/deal.II/base/tensor.h [hunk 6]
```diff
@@ -19,741 +19,213 @@
...
-  for (unsigned int i=0; i<dim; ++i)
-    subtensor[i].clear();
+  return static_cast<Number>(t) * factor;
 }
 
 
-
-template <int rank_, int dim, typename Number>
+/**
+ * Division of a tensor of rank 0 with a scalar number.
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: division...with -> division...by

File: include/deal.II/base/tensor.h [hunk 7]
```diff
@@ -19,741 +19,213 @@
...
+ * Add two tensors of rank 0.
+ *
+ * @relates Tensor<0,dim,Number>
+ */
+template <int dim, typename Number, typename OtherNumber>
 inline
-void
-Tensor<rank_,dim,Number>::serialize(Archive &ar, const unsigned int)
+Tensor<0, dim, typename ProductType<Number, OtherNumber>::type>
+operator+ (const Tensor<0,dim,Number> &p, const Tensor<0,dim,OtherNumber> &q)
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: So this is a good question -- does `ProductType` hold the appropriate data type for an addition?

File: include/deal.II/base/tensor_base.h [hunk 8]
```diff
@@ -430,79 +464,85 @@ class Tensor<1,dim,Number>
...
    * exactly it means to assign a scalar value to a tensor, zero is the only
    * value allowed for <tt>d</tt>, allowing the intuitive notation
    * <tt>t=0</tt> to reset all elements of the tensor to zero.
+   *
+   * @relates EnableIfScalar
    */
-  Tensor<1,dim,Number> &operator = (const Number d);
+  template <typename OtherNumber,
+            typename = typename EnableIfScalar<OtherNumber>::type>
+  Tensor<rank_,dim,Number> &operator = (const OtherNumber d);
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: template default arguments on functions are C++11 only. But I think you don't even need it -- we don't actually care about the data type, it simply needs to be comparable to zero, i.e., you can do
```
  Assert (d == OtherNumber(), scalar needs to be zero);
  for (i)
    data[i] = Number();   // not the 'd' given by the user, but an equally valid zero
```

File: include/deal.II/base/tensor_base.h [hunk 9]
```diff
@@ -430,79 +464,85 @@ class Tensor<1,dim,Number>
...
-   * equals the usual <tt>l<sub>2</sub></tt> norm of the vector.
+   * squares of all entries.
    */
   real_type norm () const;
 
   /**
-   * Return the square of the Frobenius-norm of a tensor, i.e. the square root
-   * of the sum of squares of all entries.
+   * Return the square of the Frobenius-norm of a tensor, i.e. the sum of
+   * squares of all entries.
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: This description is inconsistent with using `NumberTraits::abs_square()`.

File: include/deal.II/base/tensor_base.h [hunk 10]
```diff
@@ -854,28 +836,59 @@ void Tensor<0,dim,Number>::serialize(Archive &ar, const unsigned int)
...
+  // general, polymorphic for extracting an item out of an object with
+  // nested identifiers.
+  template<int rank_> struct TensorIndicesHelper
+  {
+    // used for implementing Tensor<rank,dim>::operator[] with TableIndices
+    // tail recursive call to form up access to
+    //   tensor[indices[0]][indices[1]]...[indices[rank_]]
+    template<int rank, int dim, typename Number>
+    static inline
+    Number &extract(Tensor<rank_,dim,Number> &t, const TableIndices<rank> &indices)
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: if you have a class with only a single static member function, you may as well move the function out :-)

File: include/deal.II/base/tensor_base.h [hunk 11]
```diff
@@ -1098,276 +1127,157 @@ Tensor<1,dim,Number> &Tensor<1,dim,Number>::operator /= (const Number s)
...
-Tensor<1,dim,Number>::norm_square () const
+typename Tensor<rank_,dim,Number>::real_type
+Tensor<rank_,dim,Number>::norm_square () const
 {
-  real_type s = numbers::NumberTraits<Number>::abs_square(values[0]);
-  for (unsigned int i=1; i<dim; ++i)
-    s += numbers::NumberTraits<Number>::abs_square(values[i]);
+  real_type s = 0;
+  for (unsigned int i=0; i<dim; ++i)
+    s += static_cast<Tensor<rank_-1,dim,Number> >(values[i]).norm_square();
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: do you really need the static_cast here?

File: include/deal.II/base/tensor_base.h [hunk 12]
```diff
@@ -1098,276 +1127,157 @@ Tensor<1,dim,Number> &Tensor<1,dim,Number>::operator /= (const Number s)
...
 void
-Tensor<1,dim,Number>::unroll_recursion (Vector<Number2> &result,
-                                        unsigned int    &index) const
+Tensor<rank_, dim, Number>::unroll_recursion (Vector<OtherNumber> &result,
+                                              unsigned int        &index) const
 {
   for (unsigned int i=0; i<dim; ++i)
-    result(index++) = operator[](i);
+    static_cast<Tensor<rank_ - 1, dim, Number> >(values[i]).
+    unroll_recursion(result, index);
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: same here

File: include/deal.II/base/tensor_base.h [hunk 13]
```diff
@@ -359,70 +379,84 @@ class Tensor<1,dim,Number>
...
-   * lunatic value that is likely to overflow memory limits.
+   * The tensor type this object represents.
+   */
+  typedef Tensor<rank_,dim,Number> tensor_type;
+
+  /**
+   * Type of stored objects (i.e., the object returned by operator[]()).
+   * This is a tensor of lower rank.
+   */
+  typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: I think this is simply `Tensor<rank_-1,dim,Number>` :-)

File: include/deal.II/base/tensor_base.h [hunk 14]
```diff
@@ -116,14 +115,28 @@ class Tensor<0,dim,Number>
   typedef typename numbers::NumberTraits<Number>::real_type real_type;
 
   /**
-   * Constructor. Set to zero.
+   * The tensor type this object represents. In the special case of a
+   * tensor or rank 0 we strip the tensor class and just store the scalar
+   * object.
    */
-  Tensor ();
+  typedef Number tensor_type;
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: This strikes me as a bad idea. The scalar itself is not a good representation of a tensor. I would suggest
```
  typedef Tensor<0,dim,Number> tensor_type;
```

File: include/deal.II/base/tensor_base.h (hunk 13 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: But the fact that it is not current seems to me like a bad idea.

File: include/deal.II/base/tensor_base.h (hunk 11 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: I see now why you need this, but it seems like a bad idea to me. Currently when you get down to `rank==1`, you get a scalar from `values[i]`. With the cast, you get a `Tensor<0,dim>` back which has the desired function -- but that means that when you call this function, you have to create `dim^rank` objects of type `Tensor<0,dim>`. The compiler should be able to see through this, but I think that's not the way to go. 

I think there were a few other static casts around here, which probably also have the same fundamental problem.

File: include/deal.II/base/tensor_base.h [hunk 15]
```diff
@@ -1388,163 +1298,111 @@ struct ProductType<Tensor<1,dim,T>,U>
...
-  // standard committee saw it fit to not define an
-  //   operator*(float,std::complex<double>)
-  // (as well as with switched arguments and double<->float).
-  typedef typename ProductType<Number,OtherNumber>::type product_type;
-  Tensor<1,dim,product_type> tt (false);
+  // recurse over the base objects
+  Tensor<rank,dim,typename ProductType<Number,OtherNumber>::type> tt;
   for (unsigned int d=0; d<dim; ++d)
-    tt[d] = product_type(t[d]) * product_type(factor);
+    tt[d] = t[d] * factor;
```
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: The casts were necessary since with some compilers, the product `double * std::complex<double>` is not defined :-(

File: include/deal.II/base/vectorization.h [hunk 16]
```diff
@@ -65,9 +65,19 @@ namespace std
...
 
+
+// Enable the EnableIfScalar type trait for VectorizedArray<Number> such
+// that it can be used as a Number type in Tensor<rank,dim,Number>, etc.
+
+template<typename Number>
+struct EnableIfScalar<VectorizedArray<Number> >
+{
+  typedef VectorizedArray<typename EnableIfScalar<Number>::type> type;
+};
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: nice :-)

File: include/deal.II/grid/grid_tools.h [hunk 17]
```diff
@@ -1396,7 +1396,7 @@ namespace GridTools
                 // this line has children
                 cell->face(face)->child(0)->vertex(1)
                   = (cell->face(face)->vertex(0) +
-                     cell->face(face)->vertex(1)) / 2;
+                     cell->face(face)->vertex(1)) / 2.;
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Uh, do you need this now? Surely I would expect that you can write `Point<dim>()/2` without a compiler error...

**Comment by bangerth** [comment]
Nice effort -- we didn't have `Tensor<0,dim>` when `Tensor<1,dim>` and `Tensor<rank,dim>` were written, and so it came in below the base of the recursion :-( Thanks for cleaning this up.

I added a whole bunch of annotations. Nothing that can't be addressed.

If you wanted to move everything into `tensor.h` and deprecate `tensor_base.h`, I shall be fine with this.

File: include/deal.II/base/tensor_base.h (hunk 14 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]:
```
typedef Number tensor_type;
```
Well, this is just a mechanism to preserve the old logic - namely saving the underlying data in ```Tensor<1,dim,Number>``` as a plain array.

My suggestion for a further cleanup would be to remove this mechanism altogether and always store a tensor type in ```values```. This would also remove the necessity for half of the ```static_cast<>```s.

I guess we can trust a modern compiler enough to store an array of ```Tensor<0,dim,Number>``` efficiently.

File: include/deal.II/base/tensor_base.h (hunk 13 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: No, it is not. ```value_type``` is the return type of ```operator[]()``` - at least we use it like that. :-)

If we clean up the mechanism with ```tensor_type``` described below, then we will not have to discuss anything anymore :-).

File: include/deal.II/base/tensor_base.h (hunk 11 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: As below.

I would be more than happy to remove the special notion of ```Tensor<1,dim>```s value type.

File: include/deal.II/base/tensor_base.h (hunk 10 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: Nope ;-)
This is a trick to be able to do partial template specializations on functions.

**Comment by tamiko** [comment]
> Nice effort -- we didn't have Tensor<0,dim> when Tensor<1,dim> and Tensor<rank,dim> were written, and so it came in below the base of the recursion :-(

I guessed so. I'll further clean up according to your suggestions with two major changes:
 * remove the ```tensor_type``` trick and just store tensors
 * deprecated ```tensor_base.h```

File: include/deal.II/base/tensor_base.h (hunk 10 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: Well, that's called *overloading* with functions :-)

I think you should be able to do the same with just an overload set. But it's an internal interface, so I don't particularly care whether it is done or not. I really only care about the public interfaces ;-)

File: include/deal.II/base/tensor_base.h (hunk 13 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: OK. It's a bit hard to see just from a diff, so if you are willing to clean this up so that we don't need the static casts any more, then I'm willing to let you off the hook for this patch. The issue with the static casts really bothers me enough that I think we should really make sure we don't need it.

File: include/deal.II/grid/grid_tools.h (hunk 17 above)
Comment (tamiko) [inline include/deal.II/grid/grid_tools.h]: > Uh, do you need this now? 

No, this is not needed. This is a leftover from the cleanup - I needed it for an intermediate stage.
I left those changes in because it is good style to use a floating point literal in this case.

Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Well, I think that's arguable. You're taking an average, so you're dividing by the number of elements you added up. That's an integer.

**Comment by tamiko** [comment]
@bangerth
It turns out that we really want to preserve the old logic. The reason is that we &lt;blink&gt;<i><b>need</b></i>&lt;/blink&gt; to have ```Tensor<1,dim>::operator[](unsigned int)``` to return a plain ```Number``` type.

But there is a solution to have a tensor stored internally to ground recursions:

In short the following interface works:
```
template ... class Tensor<0, dim, Number>
{
public:
  typedef Number value_type;
  operator Number &();
  operator const Number &() const; // always return a reference type!
private:
  typedef Number tensor_type;
  Number value;
};
```

```
template ... class Tensor<rank, dim, Number>
{
public:
  typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
  value_type & operator [] (const unsigned int i);
  const value_type & operator[](const unsigned int i) const;
private:
  typedef Tensor<rank_, dim, Number> tensor_type;
  Tensor<rank_-1, dim, Number> values[dim];
};
```

Now, I still have access to the underlying tensor member functions via ```values[i]``` for recursion and ```operator[]``` returns the correct stuff for external usage.

File: include/deal.II/base/point.h (hunk 3 above)
Comment (tamiko) [inline include/deal.II/base/point.h]: I will undo this. This was a well-meant mistake. ```Point``` already sports ```norm_square```.

Comment (tamiko) [inline include/deal.II/base/point.h]: And update the description anyway.

File: include/deal.II/base/tensor.h (hunk 7 above)
Comment (tamiko) [inline include/deal.II/base/tensor.h]: In general no. With C++11 I would simply write
```
auto operator+ (const Tensor<0,dim,Number> &p, const Tensor<0,dim,OtherNumber> &q)
  -> Tensor<0, dim, decltype(p[0] * q[0]);
```

But, my personal take on this is that reimplementing ```ProductType``` with a verbatim copy (~200 lines) for a single linge change in it is not worth the effort.

As soon as we drop C++98/03 support, I will remove the majority of our custom type traits and clean up places like this.

File: include/deal.II/base/tensor_base.h (hunk 8 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: The big problem is that with that it is perfectly valid to write
```
Tensor<1,dim> foo = Tensor<4,dim>();
```
I will come up with something that still allows
```
Tensor<rank,dim,complex<double>> foo = 0.;
```
to be written.

File: include/deal.II/base/tensor_base.h (hunk 15 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: libc++ shipped with XCode two years ago to be precise. I reported this upstream it should be fixed for more than two years by now. I'll give the current libc++ version a try.

**Comment by bangerth** [comment]
> It turns out that we really want to preserve the old logic. The reason is that
> we <blink>/*need*/</blink> to have |Tensor<1,dim>::operator[](unsigned int)|
> to return a plain |Number| type.

Yes, I totally agree. I didn't object to that, I simply objected to using the 
name `tensor_type` to typedef a scalar.


> But there is a solution to have a tensor stored internally to ground recursions:
>
> In short the following interface works:
>
> |template ... class Tensor<0, dim, Number>
> {
> public:
>    typedef Number value_type;
>    operator Number &();
>    operator const Number &() const; // always return a reference type!
> private:
>    typedef Number tensor_type;
>    Number value;
> };
> |
>
> |template ... class Tensor<rank, dim, Number>
> {
> public:
>    typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
>    value_type & operator [] (const unsigned int i);
>    const value_type & operator[](const unsigned int i) const;
> private:
>    typedef Tensor<rank_, dim, Number> tensor_type;
>    Tensor<rank_-1, dim, Number> values[dim];
> };
> |
>
> Now, I still have access to the underlying tensor member functions via
> |values[i]| for recursion and |operator[]| returns the correct stuff for
> external usage.
Seems sensible.

File: include/deal.II/base/tensor.h (hunk 7 above)
Comment (bangerth) [inline include/deal.II/base/tensor.h]: Yes. I didn't intend to suggest you write `AdditionType`. My idea was more to simply not allow addition in mixed types.

By the way, I think that `ProductType` is generally still the right choice if you write
```
  A a;
  B b;
  C c = a+b;
```
as
```
  A a;
  B b;
  C c = a*B(1)+b*A(1);
```
then using `ProductType` for `C` is actually quite reasonable for all "useful" number systems.

File: include/deal.II/base/tensor_base.h (hunk 8 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: Fair point. Then drop `OtherNumber` altogether and only allow `Number` as argument. The compiler will try to convert whatever you have on the right hand side to that.

File: include/deal.II/base/tensor_base.h (hunk 15 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: No, I ran into this earlier this year with GCC when helping to make the data type in `FEValues` generic (based on @davydden's patches). It's not a libc++ problem alone.

File: include/deal.II/base/tensor_base.h (hunk 10 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: In principle yes. Except: Due to type ambiguity I have to specify the recursive template parameters. I cannot do this with function parameter overloading.

File: include/deal.II/base/tensor_base.h (hunk 15 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: Which gcc version?

**Comment by tamiko** [comment]
There is one remaining issue:
```
The following tests FAILED:
        578 - base/symmetric_tensor_25.debug (Failed)
```
I will resolve this in a separate pull-request.

**Comment by tamiko** [comment]
@bangerth Found it. It was gcc-4.8 (and possibly older). I have reapplied the "workaround" for gcc versions 4.8, or older.

**Comment by tamiko** [comment]
This is now ready to be merged.

File: include/deal.II/base/tensor.h [hunk 18]
```diff
@@ -106,62 +348,80 @@ class Tensor
...
+  // array always has positive size.
 
   /**
-   * Constructor. Initialize all entries to zero.
+   * Constructor. Initialize all entries to zero if
+   * <tt>initialize==true</tt>; this is the default behaviour.
    */
-  Tensor ();
+  explicit
+  Tensor (const bool initialize = true);
```
Comment (kronbichler) [inline include/deal.II/base/tensor.h]: A question: Doesn't the construction from lower-order tensors for which the default constructor initializes the data to zero initialize the data to zero here as well, no matter what you request? I know we had this construction before, but IIRC this did not have an effect for rank >= 2. Could you please check and remove the bool argument in case it does indeed initialize to zero?

File: include/deal.II/base/tensor_base.h [hunk 19]
```diff
@@ -1,1552 +0,0 @@
-// ---------------------------------------------------------------------
```
Comment (kronbichler) [inline include/deal.II/base/tensor_base.h]: Do I read this patch correctly that tensor_base.h is completely removed? Since tensor_base.h is user-visible, don't we need to keep a dummy (with deprecation warning) for at least one release or so?

**Comment by kronbichler** [comment]
Apart from the above two comments, I'm fine with merging this patch (I guess you could remove the number of patches a bit, though).

File: include/deal.II/base/tensor_base.h (hunk 19 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: > Do I read this patch correctly that tensor_base.h is completely
> removed? Since tensor_base.h is user-visible, don't we need to keep a
> dummy (with deprecation warning) for at least one release or so?

Yes, it was user visible, but always treated as an internal header.

Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: Nah, I don't think so. I think we need to keep the header file for at least one release.

File: include/deal.II/base/tensor_base.h (hunk 15 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: I don't recall, but I think you already found out yourself. It's a pretty annoying issue.

File: include/deal.II/base/tensor_base.h (hunk 19 above)
Comment (tamiko) [inline include/deal.II/base/tensor_base.h]: Put back in place.

File: include/deal.II/base/tensor.h (hunk 18 above)
Comment (tamiko) [inline include/deal.II/base/tensor.h]: Yes, it does. We never recurse to an array over a primitive Number type, consequently, we will always initialize all values to zero via the implicit default constructors that are called.

Comment (tamiko) [inline include/deal.II/base/tensor.h]: If have removed the second recursion because it is unnecessary.

Maybe it is worthwhile to think about a more efficient allocator usage/ memset strategy in a separate pull request.

Comment (kronbichler) [inline include/deal.II/base/tensor.h]: Yes, we could think about that but I don't know how feasible it is (because it all depends on the behavior of Tensor<0,dim>. In any case, I don't think it will be overly critical because compilers tend to get better in removing initialize-to-zero code in places where it would soon be overwritten. And I guess the soon-to-be-overwritten case it the only one where not initializing the array would make sense. The other relevant case I can imagine right now is a long std::vector/AlignedVector of Tensors but there one provides an initializer element already so better copy zero than undefined memory. :-)

Comment (bangerth) [inline include/deal.II/base/tensor.h]: I think the current state is actually quite good, because compilers can decompose these arrays into individual scalars and their initialization is explicit for each element (as opposed to more difficult to figure out if you had a memset or similar that initializes *all* elements at once). The point being that if you had explicit initialization of individual scalars, the compiler can easily forward propagate this and, for example, if you happened to immediately store something in the tensor (as one often does) throw away the original initialization as a dead store. That would be much harder if you had a memset.

File: cmake/checks/check_03_compiler_bugs.cmake [hunk 20]
```diff
@@ -422,3 +422,19 @@ IF(DEAL_II_WITH_CXX11)
...
+# specialization is needed. See also base/product_type_02.cc and
+# base/product_type_03.cc
+#
+# - Matthias Maier, 2015
+#
+
+IF( CMAKE_CXX_COMPILER_ID MATCHES "GNU" AND
+    CMAKE_CXX_COMPILER_VERSION VERSION_LESS "4.9" )
+  SET(DEAL_II_GCC_COMPLEX_CONV_BUG TRUE)
+ENDIF()
```
Comment (bangerth) [inline cmake/checks/check_03_compiler_bugs.cmake]: We would need to use this in a *lot* of places in `fe_values.cc` and friends...

File: doc/news/changes.h [hunk 21]
```diff
@@ -156,9 +163,12 @@ inconvenience this causes.
   - TensorFunction
 
   <br>
-  <em>Classes now supporting complex number types via header:</em>
+  <em>Classes with fixed interface that now fully supporting complex number
```
Comment (bangerth) [inline doc/news/changes.h]: -> "that now fully support" or "that are now fully supporting"

File: include/deal.II/base/tensor.h [hunk 22]
```diff
@@ -16,14 +16,259 @@
...
+   * operator[](). This is a scalar number type for a rank 0 tensor.
+   */
+  typedef Number value_type;
+
+  /**
+   * Declare an array type which can be used to initialize an object of this
+   * type statically. In case of a a tensor of rank 0 this is just the scalar
+   * number type Number.
+   */
+  typedef Number array_type;
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: out of curiosity, would everything continue to work if you had 
```
  typedef Number array_type[1];
```
? This is not a request to change it, simply a question of curiosity.

File: include/deal.II/base/tensor.h [hunk 23]
```diff
@@ -300,153 +558,384 @@ class Tensor
...
+
+template <int dim,typename Number>
 inline
-Tensor<rank_,dim,Number>::Tensor ()
+Tensor<0,dim,Number>::Tensor ()
 {
-// default constructor. not specifying an initializer list calls
-// the default constructor of the subobjects, which initialize them
-// selves. therefore, the tensor is set to zero this way
+  value = value_type();
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: why not move this into a member initializer list as usual? same for the constructors below.

File: include/deal.II/base/tensor.h [hunk 24]
```diff
@@ -300,153 +558,384 @@ class Tensor
...
+template <int rank_, int dim, typename Number>
+inline
+Tensor<rank_,dim,Number>::Tensor (const bool /*initialize*/)
+{
+  // All members of the c-style array values are already default initialized
+  // and thus all values are already set to zero recursively.
+
+  // TODO: Think about using the default heap allocator to restore the old
+  // behavior.
+}
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: No, let's just get rid of the argument. I think it was a dumb idea anyway.

Comment (bangerth) [inline include/deal.II/base/tensor.h]: Or just deprecate the whole constructor.

File: include/deal.II/base/tensor.h [hunk 25]
```diff
@@ -300,153 +558,384 @@ class Tensor
...
+  for (unsigned int i=0; i<dim; ++i)
+    values[i] = initializer[i];
 }
 
 
+template <int rank_, int dim, typename Number>
+template <typename OtherNumber>
+inline
+Tensor<rank_,dim,Number>::operator
+Tensor<1,dim,Tensor<rank_-1,dim,OtherNumber> > () const
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: nit-picking: can you break this as
```
Tensor<rank_,dim,Number::
operator Ten... () const
```
to make it easier to read?

File: include/deal.II/base/tensor.h (hunk 22 above)
Comment (tamiko) [inline include/deal.II/base/tensor.h]: > out of curiosity, would everything continue to work if you had 
> ```
>   typedef Number array_type[1];
> ```

It would require to add another constructor and an assignment operator
for the plain Number type to Tensor<0,dim,Number>.

**Comment by bangerth** [comment]
Looks good almost everywhere. Please fix the remaining issues and merge yourself.

File: include/deal.II/base/tensor.h (hunk 22 above)
Comment (bangerth) [inline include/deal.II/base/tensor.h]: Not worth it.

File: include/deal.II/base/tensor.h (hunk 23 above)
Comment (tamiko) [inline include/deal.II/base/tensor.h]: Ask the author who wrote this 10 years ago ;-)

File: include/deal.II/base/tensor.h (hunk 24 above)
Comment (tamiko) [inline include/deal.II/base/tensor.h]: This would will also imply to do the same change for Point<dim,Number>.

Comment (tamiko) [inline include/deal.II/base/tensor.h]: I will remove it. I doubt that this boolean is used terribly often in user code.

Comment (bangerth) [inline include/deal.II/base/tensor.h]: I agree. Do the same for `Point<dim>`
